mne/io/eyetracking/utils.py

```diff
@@ -0,0 +1,265 @@
+# Author: Dominik Welke <dominik.welke@web.de>
+#
+# License: BSD-3-Clause
+
+import numpy as np
+from ..constants import FIFF
+from ...utils import warn
+
+
+def set_channelinfo_eyetrack(inst, mapping=None,
+                             overwrite=False, verbose=None):
+    """
+    updates channel information for eyetrack channels
+    (if overwrite=True, all channels that are not recognized as proper eyetrack
+    channels will be set to MISC)
+    this handles the deeper implementation using FIFF types
+
+    user can either rely on automatic extraction from channel name (str), or
+    provide a dictionary with entries for each channel name.
+
+    example:
+    mapping = {
+        "L-GAZE-X": {
+            "ch_type": "GAZE",
+            "unit": "PIX",
+            "eye": "L",
+            "xy": "X"
+            },
+        "R-AREA": {
+            "ch_type": "PUPIL",
+            "unit": "AU",
+            "eye": "R"
+            }
+        }
+    """
+    inst.info = _set_fiff_channelinfo_eyetrack(
+        inst.info, mapping=mapping, overwrite=overwrite, verbose=verbose)
+    return inst
+
+
+def _eyetrack_channelinfo_from_chname(ch_name):
+    """
+    guess specs of the eyetrack channel from ch_name:
+    ch_type, unit, eye (l/r), axis (x/y)
+    """
+
+    # try improving detection by splitting the channel name
+    seplist = ['-', '_', '/']  # assume only one is present
+    ch_parts = []
+    for sep in seplist:
+        if sep in ch_name:
+            ch_parts = ch_name.lower().split(sep)
+
+    # extract info
+    ch_type = (
+        'eyetrack_pupil' if (
+            'pupil' in ch_parts or
+            'area' in ch_parts or
+            'pupil' in ch_name.lower() or
+            'area' in ch_name.lower()
```

Comment:
So as far as the  I/O (in this branch) is concerned, the channel names are predetermined by `read_raw_eyelink`. A pupil channel will always be `pupil_left` or `pupil_right`.  Likewise, information about the channel type (i.e. that  _it is_ a pupil channel, and whether pupil _diameter_ or pupil _area_ was output) is always available in the ASCII file and is (or can be) parsed by `read_raw_eyelink`.  My point is that we don't have to guess what the channel type or channel info _could_ be based off of what the channel name _might_ be.

If you intended for this function to have a sort of general purpose utility for setting channel info for eyetracking channels broadly (i.e., if you intend for it to be used in the future if a Tobii I/O is implemented) - then I think that it would be better placed in `mne.io.utils`, the same way that `mne.io.utils._create_chs` is used to   _"initialize info['chs'] for eeg channels."_ (I'm taking that docstring descprition at face value).

Reply:
yes, of course this is intended to be general purpose: besides eyelink and tobii there are plenty other systems, but importantly also co-registered recordings (e.g. via the matlab eye-eeg toolbox or lsl) that can already be loaded in MNE. 
so it is sensitive for us to provide a way to set full channel specs without needing a specific reader function for all cases.

i dont mind this place for it: it is specific to eyetracking and it is in a general `util` file, not in your `eyelink` file. yet, it's also easy to move somewhere else. the reviewers can tell us where they think it makes most sense

regarding the eyelink reader: ok sure, if you are certain that the .asc files *really* always come with the same channel names, we can change it to a hard coded version. with SR being there forever doing their proprietary thing this might even be the case.. but honestly, there is much more inconsistency than you should expect, even with seemingly standardized file formats - maybe someone wants to revisit a file from 10 years ago, generated with an outdated version of the file format? ;)  


btw, did you hear back from SR regarding the measurement date?
this value is so far not set in the `info`

Reply:
> i dont mind this place for it: it is specific to eyetracking and it is in a general `util` file, not in your `eyelink` file. yet, it's also easy to move somewhere else. the reviewers can tell us where they think it makes most sense

This is something I wanted to bring up actually.. :grin:. I _really_ don't think that this sub-package should be called `mne.io.eyetracking`. It's inconsistent with current the sub-package organization in `mne.io`, where all other sub-packages are organized by manufacturer. There is no `fnirs` sub-package, rather there are sub-packages for `nirx` and `snirf`. The same is true for MEG and EEG. 

If you agree, then I would suggest `mne.io.eyelink` and this is why I also suggest that more general purpose functions don't reside in this sub-package.

I'm sorry if I sound pedantic or hand-wavy, but I bring it up because I don't think its a trivial matter. Then again, you started this work during the code sprint with the core-devs involved so if others think that this sub-package should be `mne.io.eyetracking`, with any manufacturer specific readers (Eyelink, tobii, others) living inside this subpackgage - I can live with being the odd on out here :slightly_smiling_face:

Reply:
> I _really_ don't think that this sub-package should be called `mne.io.eyetracking`. It's inconsistent with current the sub-package organization in `mne.io`, where all other sub-packages are organized by manufacturer. [...] If you agree, then I would suggest `mne.io.eyelink` and this is why I also suggest that more general purpose functions don't reside in this sub-package.

That sounds correct to me. @larsoner I think you worked most closely with @dominikwelke during the sprint so presumably you were at one point OK with `mne.io.eyetracking`... do you have an opinion here?

(aside: I have some Eyelink files recorded in 2013, I will email one to both of you for testing)

Reply:
Responding to Dominiks other questions/points that I didn't before:

> btw, did you hear back from SR regarding the measurement date?
> this value is so far not set in the `info`

Thank you for reminding me! I think it's in my inbox somewhere.. I'll follow up!

>  honestly, there is much more inconsistency than you should expect, even with seemingly standardized file formats

Yes I take your point on that. Well, If we take the SR Manual on their ASCII specification at face value (and confirm this via testing of a few test files), yes we should know the column labels for the columns that are present, and their order.. As long as the user accepts the default parameters for ASCII conversion (i.e. they dont make eyellink literally just output samples without any corresponding info on the data in the ASCII file). Of course there can always be bugs on eyelink's side that cause deviations from their own specification, but IMO we should tackle those if/when they arise.

As for versioning.. Ideally Eyelink would include release notes if they update their file format. Then if needed, we have code to handle file differences inside the `eyelink` sub-package, the same way that the `mne.io.EGI` does for EGI .mff files vs the old EGI binary format..

> co-registered recordings (e.g. via the matlab eye-eeg toolbox or lsl) that can already be loaded in MNE.
> so it is sensitive for us to provide a way to set full channel specs without needing a specific reader function for all cases.

I didn't know that MNE main could load eyetracking data from the matlab eye-eeg toolbox, that's cool! I am not against having a general utility function, just a suggestion on where it should reside.

Reply:
> This is something I wanted to bring up actually.. grin. I really don't think that this sub-package should be called mne.io.eyetracking. It's inconsistent with current the sub-package organization in mne.io, where all other sub-packages are organized by manufacturer. There is no fnirs sub-package, rather there are sub-packages for nirx and snirf. The same is true for MEG and EEG.

I think all that really needs to be decided at this point is the public API endpoint(s). I'd expect this to be `mne.io.read_raw_eyelink` whose code lives in `mne/io/eyelink/_*.py`, i.e., nested according to the manufacturer. We can also have a `mne/io/_eyetracking.py` if there are enough eyetracking-related utilities expected to warrant its own file in the `io` root. But note the leading underscores to keep it private -- this means we can move/rename whenever we want if we decide this organization isn't ideal

Reply:
Okay, that makes sense, and @dominikwelke sorry getting sidetracked on other tangents there.. I do see your point about using utility functions to limit code redundancy. Per @larsoner 's suggestions I'll let you handle the organization of those as you see fit.

Reply:
no worries at all @scott-huberty, i agree that the reader module should be called eyelink didnt intend to argue about that :)

i guess i was just less concerned, because renaming folders, files and __init__ entries is less work then the actual functionality and can be done anytime before merging. 

i just started the PR, but by now youve certainly worked more on it than me!
so feel free to rename the submodule and re-replace the general purpose function in `_create_info()` with a hard coded version.
after that i'll shift the ch_name parser defs somewhere else.

Reply:
sweet! 

btw, i think you said that you have the units in the asci files..
it would be great to set these as well, as some easy and handy preprocessing methods would be unit converter (px<->deg, au<->mm)

on a related note: 
it *might* make sense to add certain other metainfo, e.g. the reference point for gaze position (screen corner vs center)
whats your opinion, is there anything else in the header that's so important we should definitely record it?

Reply:
> btw, i think you said that you have the units in the asci files..

in the ASCII files, lines that start with SAMPLES indicate the eyepos unit. The keyword after SAMPLES is GAZE/HREF/PUPIL for pixels/visual degrees/raw coordinates respectively. Lines that start with PUPIL indicate pupil datatype. the keyword after PUPIL is DIAMETER/AREA for pupil diameter or pupil area.

We store the samples unit info in the private attribute `_.event_lines['SAMPLES']`. Right now we are not storing the pupil info but if  add `'PUPIL'` to  the `dict` on line 398 of `eyelink.py`, this info will be captured.



> it _might_ make sense to add certain other metainfo, e.g. the reference point for gaze position (screen corner vs center)
> whats your opinion, is there anything else in the header that's so important we should definitely record it?

Yeah, there's more info on the tracking session that could be gleaned. I started writing code to parse this a while back, but it quickly got complex so I kept it off this branch, with the idea that it could be added in a follow up PR if worth it. That code is on a branch named `mne-eyetrack_metainfo` on my fork, if you ever want to see what information gets returned there so far.